### PR TITLE
Add new audit handler method for action responses

### PR DESCRIPTION
**Pull request by albertzaharovits**
This adds a new method to the `AuditTrail` that intercepts the responses of transport actions. This new method is unlike all the other existing audit methods because it is called after the action has been run (so that it has access to the response).

In order to facilitate the "linking" of the new audit event with the other existing events, the audit method receives the `requestId` and the `authentication` as arguments (in addition to the `request` itself and the `response`).  

This is labeled _non-issue_ because it is only the foundation upon which later features that actually print out (some) responses can be built upon.

Related https://github.com/elastic/elasticsearch/pull/63221

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java [hunk 1]
```diff
@@ -47,6 +48,10 @@ void accessGranted(String requestId, Authentication authentication, String actio
     void accessDenied(String requestId, Authentication authentication, String action, TransportRequest transportRequest,
                       AuthorizationInfo authorizationInfo);
 
+    // this is the only audit method that is called *after* the action executed, when the response is available
+    void actionResponse(String requestId, Authentication authentication, String action, TransportRequest transportRequest,
+                        TransportResponse transportResponse);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:53]: the new audit method.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java [hunk 2]
```diff
@@ -83,25 +86,24 @@ public SecurityActionFilter(AuthenticationService authcService, AuthorizationSer
         if (licenseState.isSecurityEnabled()) {
             final ActionListener<Response> contextPreservingListener =
                     ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
+            final ActionListener<Response> postActionExecutionListener = ActionListener.delegateFailure(contextPreservingListener,
+                    (ignore, response) -> {
+                        String requestId = AuditUtil.extractRequestId(threadContext);
+                        Authentication authentication = securityContext.getAuthentication();
+                        auditTrailService.get().actionResponse(requestId, authentication, action, request, response);
+                        contextPreservingListener.onResponse(response);
+                    });
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:95]: This intercepts the response for the action that is executed inside the rest handler (that do NOT go to other cluster nodes).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 3]
```diff
@@ -225,7 +237,36 @@ public void onFailure(Exception e) {
...
+                            Authentication authentication = securityContext.getAuthentication();
+                            auditTrailService.get().actionResponse(requestId, authentication, action, request, response);
+                            channel.sendResponse(response);
+                        }
+
+                        @Override
+                        public void sendResponse(Exception exception) throws IOException {
+                            channel.sendResponse(exception);
+                        }
+                    }, task);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:269]: This intercepts the response for actions executed on other cluster nodes (when an action on the coordinating node submits requests to other nodes - or even the same node)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java [hunk 4]
```diff
@@ -83,25 +86,24 @@ public SecurityActionFilter(AuthenticationService authcService, AuthorizationSer
         if (licenseState.isSecurityEnabled()) {
             final ActionListener<Response> contextPreservingListener =
                     ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
+            final ActionListener<Response> postActionExecutionListener = ActionListener.delegateFailure(contextPreservingListener,
+                    (ignore, response) -> {
+                        String requestId = AuditUtil.extractRequestId(threadContext);
+                        Authentication authentication = securityContext.getAuthentication();
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:92]: I haven't mentioned this as a drawback when we discussed it. I think I wasn't fully confident in my ability to clearly explain it. But this is part of the complexity and maintenance burden that I've mentioned, so I'll try to explain it here:

In order to pass the `requestId` and `authentication` as arguments to the new audit method, they need to be pulled from the thread context. Currently the `requestId` and `authentication` ARE present in the thread context when the listener is called with the response. But to ensure this forward we'll have to introduce an assertion (probably at this very place in the code). Even then, we cannot guarantee it because tests don't cover all the paths in an action handler: it could be that some handler stashes the thread context and calls the listener with a response (without calling another transport action, which is the usual course of action when stashing the context).

Alternatively, we could enforce it more stringently. Instead of an assert we could hard fail (reroute the call to `listener.onResponse` to `listener.onFailure`) if we detect that the thread context has been altered.

I favour the latter, but maybe this is overdoing it since the "danger" is inaccurate (missing `requestId` and `authentication`) audit logging for action responses (for which there is not concrete use case atm, but think of auditing creating API keys as an example).

We could also opt for assets in the minor and hard fail in the major release, similar to what we've discussed for indices privileges.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:97]: > Alternatively, we could enforce it more stringently. Instead of an assert we could hard fail (reroute the call to listener.onResponse to listener.onFailure) if we detect that the thread context has been altered.

I've pushed https://github.com/elastic/elasticsearch/pull/63708/commits/2f5e54682ab7abfb8f8e4c47d9b1dd5f70ab8503
 to better illustrate what I mean.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java]: I've iterated on the issue, this is no longer a problem.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java [hunk 5]
```diff
@@ -297,17 +299,26 @@ public void testTokenMissing() throws Exception {
 
             Mockito.doReturn(List.of(secondRealm)).when(realms).getUnlicensedRealms();
             Mockito.doReturn(List.of(firstRealm)).when(realms).asList();
-            final String reqId = AuditUtil.getOrGenerateRequestId(threadContext);
+            boolean requestIdAlreadyPresent = randomBoolean();
+            SetOnce<String> reqId = new SetOnce<>();
+            if (requestIdAlreadyPresent) {
+                reqId.set(AuditUtil.getOrGenerateRequestId(threadContext));
+            }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java:306]: Tests in this class have been modified to ensure that a `requestId` is set when done.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceTests.java [hunk 6]
```diff
@@ -274,7 +276,8 @@ public void setup() {
             null, Collections.emptySet(), licenseState, new IndexNameExpressionResolver(new ThreadContext(Settings.EMPTY)));
     }
 
-    private void authorize(Authentication authentication, String action, TransportRequest request) {
+    private void authorize(Authentication authentication, String action, TransportRequest request,
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceTests.java:279]: Similarly to the changes in the `AuthenticationServiceTests`, the tests in this class have been modified to assert that the `requestId` is set following a successful authz.

**Review by ywangd — COMMENTED** [review]
I took an initial look and have a few questions:
* Are we aiming to have a generic response auditing system or is it just for security changes? I assume it is a generic solution? As a generic solution, I wonder whether we would consider auditing the response on the requesting side (when it handles the response) instead of on the responding side (when it sends the response)?
* Do we want to audit errors as well? I remember we talked about this briefly and the consensus seems to be "yes"? Or do you consider the lack of response is an indication of error?
* We could be double auditing with audit in both `SecurityServerTransportInterceptor` and `SecurityActionFilter`. The former often leads to the later in the same request. I guess it could have values if the credentials are changed in the later so that two logging entries will have some difference because of different authentication. But it will probably still have lots of repetition.
* The `postAuthzListener` is rather involved. A main reason is due to the scope mismatch between auth info and response. I don't really have any good suggestions. One thing I wonder  is whether we could split part of it into `SecurityActionFilter#authorizeRequest`. The `authentication` object is avilable there so it can be passed and save a `null` check. But it needs a few more changes to work, e.g. change the `Void` listener to `Authentication`. So not sure whether it is worthy.

I will take a closer look later. But I hope above helps to get the discussion started. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java [hunk 7]
```diff
@@ -188,14 +188,15 @@ public void authorize(final Authentication authentication, final String action,
             if (auditId == null) {
                 // We would like to assert that there is an existing request-id, but if this is a system action, then that might not be
                 // true because the request-id is generated during authentication
-                if (isInternalUser(authentication.getUser()) != false) {
+                if (isInternalUser(authentication.getUser())) {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:191]: We don't ;)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java [hunk 8]
```diff
@@ -188,14 +188,15 @@ public void authorize(final Authentication authentication, final String action,
...
                     auditId = AuditUtil.getOrGenerateRequestId(threadContext);
                 } else {
                     auditTrailService.get().tamperedRequest(null, authentication, action, originalRequest);
                     final String message = "Attempt to authorize action [" + action + "] for [" + authentication.getUser().principal()
                             + "] without an existing request-id";
                     assert false : message;
                     listener.onFailure(new ElasticsearchSecurityException(message));
+                    return;
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:199]: yes

**Comment by albertzaharovits** [comment]
Thank you for looking into it, Yang!

>Are we aiming to have a generic response auditing system or is it just for security changes? I assume it is a generic solution? As a generic solution, I wonder whether we would consider auditing the response on the requesting side (when it handles the response) instead of on the responding side (when it sends the response)?

A generic solution indeed. Generic in the sense that the new audit method is generic, but the implementation can be specific for response types.
The principle that I've used to pick between the two is that given that we audit the request on the "data" node, then this should audit the response on the "data" node too; keep them together.

> Do we want to audit errors as well? I remember we talked about this briefly and the consensus seems to be "yes"? Or do you consider the lack of response is an indication of error?

I don't foresee a use case for auditing error responses; only success responses. It might be useful, though. It should be easy to add it when/if we need it (i.e. let's defer it for a follow-up).

> We could be double auditing with audit in both SecurityServerTransportInterceptor and SecurityActionFilter. The former often leads to the later in the same request. I guess it could have values if the credentials are changed in the later so that two logging entries will have some difference because of different authentication. But it will probably still have lots of repetition.

I don't believe this is the case. It should not cause double auditing. One audits the responses of `TcpTransport#inboundHandler`s and the other of `ActionModule#actions` ; different handlers different responses.

> The postAuthzListener is rather involved. A main reason is due to the scope mismatch between auth info and response. I don't really have any good suggestions. One thing I wonder is whether we could split part of it into SecurityActionFilter#authorizeRequest. The authentication object is avilable there so it can be passed and save a null check. But it needs a few more changes to work, e.g. change the Void listener to Authentication. So not sure whether it is worthy.

Your suggestion is a bit vague, so it's hard to compare against. But I think it's worth pointing out that the listener to the `chain.proceed(...` must be wrapped.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java [hunk 9]
```diff
@@ -83,29 +88,38 @@ public SecurityActionFilter(AuthenticationService authcService, AuthorizationSer
         if (licenseState.isSecurityEnabled()) {
             final ActionListener<Response> contextPreservingListener =
                     ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
-            ActionListener<Void> authenticatedListener = ActionListener.wrap(
-                    (aVoid) -> chain.proceed(task, action, request, contextPreservingListener), contextPreservingListener::onFailure);
+            final ActionListener<Void> postAuthzListener = ActionListener.delegateFailure(contextPreservingListener,
+                    (ignore, aVoid) -> {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:92]: > it feels abit strange to have (ignore, aVoid)

Sure, but the alternative of having variables similarly named pointing to the same thing is not preferable IMO.

> Alternatively, we could just use ActionListener#wrap if we do wanna ignore the delegated listener.

I don't know about this either. I think that If I were to come up with the `ActionListener#wrap` in the first place, I might be getting the converse suggestion (no flame).

**Comment by ywangd** [comment]
> I don't believe this is the case. It should not cause double auditing. One audits the responses of TcpTransport#inboundHandlers and the other of ActionModule#actions ; different handlers different responses.
>
This depends on how a `TransportRequestHandler` is registered. For subclasses of `TransportReplicationAction`, e.g. `TransportShardBulkAction`, the request handlers are registered without having to go through the action filters. But for subclasses of `HandledTransportAction` (lots of them), e.g. `TransportDeleteIndexAction`, it will always go through action filters. So using `DELETE index` as an example, if the request is issued to against a non-elected-master node, it will be forwarded to the master node, where it will first go through the inboundHandler, then go through action filters and gets audited twice. In addition, the request will be audited (in action filter) on the non-elected-master node as well. So it will be audited 3 times in total. This is one of the reasons why I wonder whether auditing the response on the requesting side is better because it could have 1 less repetition.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:191]: Right, sorry I thought it was `==`.

**Comment by albertzaharovits** [comment]
>  But for subclasses of HandledTransportAction (lots of them), e.g. TransportDeleteIndexAction, it will always go through action filters.

Looks like you're right. `HandledTransportActions` are intercepted by both the `ServerTransportFilter` and the `SecurityActionFilter`. Consequently there is double auditing. **Good catch!**
However, there is double auditing for `access_granted` events as well, not only for this newly introduced event. The problem is not in this PR. This follows the pattern of the other audits.

My suggestion is tackle the double auditing problem as a separate issue. Would you agree @ywangd ?

**Comment by ywangd** [comment]
> However, there is double auditing for `access_granted` events as well, not only for this newly introduced event. The problem is not in this PR. This follows the pattern of the other audits.
> 
> My suggestion is tackle the double auditing problem as a separate issue. 
>
True. The duplication is not unique to the response. I agree we can tackle it as a separate issue. Since users often complaint about the verbosity of audit logs and both `access_grant` and the current WIP `response` loggings are quite heavy in volume, it might be in our interest to address it sooner.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:92]: This is not a big issue. So I am happy to let it slide.

**Comment by ywangd** [comment]
>> The postAuthzListener is rather involved. A main reason is due to the scope mismatch between auth info and response. I don't really have any good suggestions. One thing I wonder is whether we could split part of it into SecurityActionFilter#authorizeRequest. The authentication object is avilable there so it can be passed and save a null check. But it needs a few more changes to work, e.g. change the Void listener to Authentication. So not sure whether it is worthy.
>>
> Your suggestion is a bit vague, so it's hard to compare against. But I think it's worth pointing out that the listener to the chain.proceed(... must be wrapped.
>

My main point is whether `postAuthzListener` (previously `authenticatedListener`) could be simplified a bit. The idea was to move creation of `postAuthzListener` into `SecurityActionFilter#authorizeRequest` where the `authentication` object is available (so no need to retrieve it from securityContext and performs a `null` check) and is the place where `postAuthzListener` is truly used. But as I said in the original comment, it incurs quite some changes. So may not worth it in terms of complexity.

**Comment by albertzaharovits** [comment]
> The idea was to move creation of postAuthzListener into SecurityActionFilter#authorizeRequest where the authentication object is available (so no need to retrieve it from securityContext and performs a null check) and is the place where postAuthzListener is truly used.

Okay, I've pushed a730658647ddcaaad3f3159f127e9b37312be40d  . Because it doesn't require accessing the thread context I think it is an improvement, thank you.

**Review by ywangd — COMMENTED** [review]
Do we aim to merge this PR for 7.11? I am still torned about having auditing in `SecurityServerTransportInterceptor`. On one hand it has great coverage. On the other hand, it does a bit too much. I wonder whether it is acceptable to just audit within `SecurityActionFilter` because the response from `SecurityServerTransportInterceptor` will be recieved by a listener which itself sits inside a `SecurityActionFilter`(either on the same node or a different node) and will have its overall response audited. In another word, whether just auditing on the sender side is a possible solution.

**Comment by albertzaharovits** [comment]
> Do we aim to merge this PR for 7.11? 

This PR introduces the audit hook method that is going to be used to audit transport responses. This is currently needed to be able to intercept API key ids in https://github.com/elastic/elasticsearch/pull/62916 .
Generally, the version we're aiming a feature for is not part of the decision to merge a PR or not.

> I am still torned about having auditing in SecurityServerTransportInterceptor. On one hand it has great coverage. On the other hand, it does a bit too much. I wonder whether it is acceptable to just audit within SecurityActionFilter because the response from SecurityServerTransportInterceptor will be recieved by a listener which itself sits inside a SecurityActionFilter(either on the same node or a different node) and will have its overall response audited. In another word, whether just auditing on the sender side is a possible solution.

I don't think only intercepting responses from the `SecurityActionFilter` is satisfactory. It leaves un-audited many responses. I seriously question the usefulness of a method used to audit only some responses. I would prefer a method that audits all responses, some of which twice.

That being said, I've spent today trying to figure out a way to prevent the doubling situation.
My assessment is that the best way to prevent it is to skip authorizing the same request on the same action, in a row, on the same thread context. It is similar to how authentication is not repeated in the same thread context, but in principle it makes more sense to preserve the authentication status than to preserve the authorization of the last request. 

Ideas welcome.

**Comment by ywangd** [comment]
>> On one hand it has great coverage. On the other hand, it does a bit too much. I wonder whether it is acceptable to just audit within SecurityActionFilter because the response from SecurityServerTransportInterceptor will be recieved by a listener which itself sits inside a SecurityActionFilter(either on the same node or a different node) and will have its overall response audited. In another word, whether just auditing on the sender side is a possible solution.
>>
> I don't think only intercepting responses from the SecurityActionFilter is satisfactory. It leaves un-audited many responses. I seriously question the usefulness of a method used to audit only some responses. I would prefer a method that audits all responses, some of which twice.
>
I think I have not been clear enough in my comment. Let me try again. 

This is _not_ about the "double auditing" that we discussed before. It is about whether we could just audit purely on the sender side (coordinating node). Any requests handled by `SecurityServerTransportInterceptor` must be sent from a `transportService` that is itself within a `SecurityActionFilter`. For an example, any transport action that is a subclass of `TransportNodesAction` will fan out sub-requests to all the nodes. These sub-requests pass through the `SecurityServerTransportInterceptor` on each node and their responses are audited. These responses then get sent back to the coordinating node where the sub-requests are sent from. At the coordinating node, the responses are aggregated into the final response. This final response is then audited again because it's inside a `SecurityActionFilter`. If we recall that the responses can already be double audited on the remote nodes, now this becomes "triple audited" because the flow is `SecurityActionFilter -> SecurityServerTransportInterceptor -> SecurityActionFilter`. It is true that the aggregated response may not be as detailed as the original responses generated by each node. So there are some values in auditting all of them. However, this value diminishes when the action is a `TransportMasterNodeAction` where the response from a remote node is basically unchanged at the coordinating node. 

In summary, auditing only on the sender side (coordinating node) should technically cover everything with the caveat that it may not have every details due to aggregation. With that being said, I recognise the value of auditing absolutely everything with no short of details. That's why I am torned as stated in the previous comment.

**Comment by albertzaharovits** [comment]
I think I understand it, Yang.

You're describing that responses of child actions don't need to have their responses audited, because those responses are collected by the parent action that itself emits a response that is audited. Please correct me if I'm wrong.

This is sensible, but the argument that I'm maintaining is that we don't have the concept of auditing actions on the coordinating node only. Until now all actions are audited at the place that they're handled. The response audit method will be the sole one that is called selectively on actions, ie some actions will not have their response audited, but they're authorization might be audited. More concretely, the log hook method that we're adding will be called only for actions that run on the coordinating node, so we need to come up with an ingenious name. (Maybe we can come up with **two** log hook methods, one for the action filter one for the transport filter, auditing incurs a lot of internal knowledge anyway.)

Though I see the merit in your proposal because it would obviate the need to skip the authorization, as described above, which I expect to be an even longer debate. It would also be sufficient for the immediate need to intercept the API Key ids.

I am going to ask this be discussed more broadly during the team meeting.

For reference, the child actions are: all actions with the `[n]`, `[p]`, or `[r]` suffix, internal actions, and some of the actions with the `[s]` suffix, plus all the following actions:
"indices:data/read/search[free_context/scroll]"
"indices:data/read/search[free_context]"
"indices:data/read/search[clear_scroll_contexts]"
"indices:data/read/search[phase/dfs]"
"indices:data/read/search[phase/query]"
"indices:data/read/search[phase/query/id]"
"indices:data/read/search[phase/query/scroll]"
"indices:data/read/search[phase/query+fetch/scroll]"
"indices:data/read/search[phase/fetch/id/scroll]"
"indices:data/read/search[phase/fetch/id]"
"indices:data/read/search[can_match]"
"indices:data/read/open_reader_context"

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java [hunk 10]
```diff
@@ -81,4 +82,9 @@ void runAsDenied(String requestId, Authentication authentication, RestRequest re
     void explicitIndexAccessEvent(String requestId, AuditLevel eventType, Authentication authentication, String action, String indices,
                                   String requestName, TransportAddress remoteAddress, AuthorizationInfo authorizationInfo);
 
+    // this is the only audit method that is called *after* the action executed, when the response is available
+    // it is however *only called for coordinating actions*, which are the actions that a client invokes as opposed to
+    // the actions that a node invokes in order to service a client request
+    void coordinatingActionResponse(String requestId, Authentication authentication, String action, TransportRequest transportRequest,
+                                    TransportResponse transportResponse);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:89]: This is not technically true, some actions' (`TransportMasterNodeAction`) responses are intercepted on the coordinating node, as well as on the master node.

The naming and descriptions are difficult to write to convey the subset of actions that are covered.
We could also bleed some internals in here and mention the `SecurityActionFilter`.

Open to suggestions.

**Comment by albertzaharovits** [comment]
@ywangd As convened, I have chopped this PR to insert the new audit hook method only on the responses of actions intercepted by the `SecurityActionFilter`.

Ready for feedback.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:89]: I think we could keep the previous name, i.e. `actionResponse`. Please let me explain: When I suggested pulling out the code from `SecurityServerTransportInterceptor`, it was in the context of getting auditing for API key ID in for v7.11.0. The reasons are:
* The response in `SecurityServerTransportInterceptor` is a fairly low level indexing response and does not really provide extra valuable information than the one in the coordinating node. Therefore auditing on just the coordinating node is sufficient.
* Creating API key is underlyingly a `TransportReplicationAction`. Unlike `TransportMasterNodeAction`, it does _not_ go through the `SecurityActionFilter` (on the remote node). This means: on the remote node, we will not get another auditing entry from `SecurityActionFilter`, i.e. there will only be a single auditing entry about API key creation and it is on the coordinating node.

So my motivation was rather narrowly focused on the particular use case. The agreement, as I understand it, is to defer the auditing code in `SecurityServerTransportInterceptor` until the next release. Because it is not necessary for API key auditing in this release and if it is not in use, it's better to be left out. But we also agreed that it could still be useful when we expand the usage scenario and this is subject to further discussion. 

IIUC, your intention of changing the method name is for the possibility to have two different audit event types. I should have been clear that it was not my proposal. I consider whether auditing inside `SecurityServerTransportInterceptor` an implementation detail, which we chose to opt out for initial release, but could choose to add it later as an enhancement. For verbosity and duplication control, I'd prefer to solve it in a different way other than splitting it into different audit event types. I hope this makes sense.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java [hunk 11]
```diff
@@ -331,10 +342,20 @@ public void testAuthenticateBothSupportSecondSucceeds() throws Exception {
         } else {
             when(secondRealm.token(threadContext)).thenReturn(token);
         }
-        final String reqId = AuditUtil.getOrGenerateRequestId(threadContext);
+        boolean requestIdAlreadyPresent = randomBoolean();
+        SetOnce<String> reqId = new SetOnce<>();
+        if (requestIdAlreadyPresent) {
+            reqId.set(AuditUtil.getOrGenerateRequestId(threadContext));
+            assertThat(reqId.get(), not(nullValue()));
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/AuthenticationServiceTests.java:349]: Nit: this assertion of reqId.get() being non-null does not exist in other places. Do we need it here?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java (hunk 10 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:89]: Thanks for the detailed explanation Yang, but I would like to focus the conversation on the scope of this PR only. We should consider the broader picture, as you describe, but ultimately this PR is what gets merged as a unit, and what we merge in the future is subject to debate still.

If we merge this PR, which adds the audit hook for the responses of actions intercepted by the `SecurityActionFilter` only, we add to the interface of the `LoggingAuditTrail` .
When we later merge the code to allow intercepting the responses of the other actions, we have the options to:
* reuse the same `LoggingAuditTrail` hook method
* rename the existing hook method, or
* add a new audit response hook method.
(or any mix of them, really, if we feel wild)

Because this is a private interface, I don't believe we have to name it in anticipation of its use. We should name it for the purpose that it serves right now, as we can always tinker with it later. For this reason, I would not favor the `actionResponse` name.

**Review by ywangd — APPROVED** [review]
LGTM. Thanks for the iterations!

Other than the following few minor comment. I also commented on the method name of the new audit method. The changes to AuthenticationServiceTests are quite noisy, but I cannot think of a better way of doing it without major revamping. I think I won’t bother with it given the changes do the right job and they are test only code.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java [hunk 12]
```diff
@@ -130,13 +124,13 @@ public int order() {
         return Integer.MIN_VALUE;
     }
 
-    private <Request extends ActionRequest> void applyInternal(String action, Request request,
-                                                               ActionListener<Void> listener) throws IOException {
+    private <Request extends ActionRequest, Response extends ActionResponse> void applyInternal(Task task, ActionFilterChain<Request,
+            Response> chain, String action, Request request, ActionListener<Response> listener) {
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:128]:
```suggestion
    private <Request extends ActionRequest, Response extends ActionResponse> void applyInternal(Task task,
            ActionFilterChain<Request, Response> chain, String action, Request request, ActionListener<Response> listener) {
```

**Review by tvernum — APPROVED** [review]

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:89]: Just trying to brainstorm here: How about calling this one `actionResponse` and the other one `transportResponse`?
As you said, this is a private interface, so we don't have to struggle too much with its name now. So please feel free to use whatever works for you.

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java (hunk 10 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:89]: Yeah, this could work. I'll keep this in mind when we'll iterate on the interface.
